File: openfilz-sdk/openfilz-sdk-csharp/samples/Quickstart.cs
// OpenFilz C# SDK Quick Start
//
// Demonstrates the core document management operations:
// - Folder creation, listing
// - File upload, download, rename, move, copy, delete
// - Favorites, metadata, dashboard, audit trail
//
// Prerequisites:
//   - A running OpenFilz API instance (default: http://localhost:8081)
//   - dotnet add package OpenFilz.Sdk
//
// The C# SDK uses Microsoft Dependency Injection (generichost pattern).
// All API classes are resolved from IServiceProvider and methods return
// IXxxApiResponse wrappers — use .Ok() to access the response data.

using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.Hosting;
using OpenFilz.Sdk.Api;
using OpenFilz.Sdk.Client;
using OpenFilz.Sdk.Extensions;
using OpenFilz.Sdk.Model;

namespace OpenFilzSdkSamples;

public class Quickstart
{
    public static async Task RunAsync(string? apiUrl = null)
    {
        apiUrl ??= Environment.GetEnvironmentVariable("OPENFILZ_API_URL") ?? "http://localhost:8081";

        // ──── 1. Configure the SDK client via HostBuilder ───────────────────
        var host = Host.CreateDefaultBuilder()
            .ConfigureApi((context, services, options) =>
            {
                // For authenticated instances:
                // options.AddTokens(new BearerToken("your-oauth2-access-token"));

                options.AddApiHttpClients(client =>
                {
                    client.BaseAddress = new Uri(apiUrl);
                });
            })
            .Build();

        var folderApi = host.Services.GetRequiredService<IFolderControllerApi>();
        var documentApi = host.Services.GetRequiredService<IDocumentControllerApi>();
        var fileApi = host.Services.GetRequiredService<IFileControllerApi>();
        var favApi = host.Services.GetRequiredService<IFavoritesApi>();
        var dashboardApi = host.Services.GetRequiredService<IDashboardApi>();
        var auditApi = host.Services.GetRequiredService<IAuditControllerApi>();

        // ──── 2. Create a folder ────────────────────────────────────────────
        var folderResp = await folderApi.CreateFolderAsync(
            new CreateFolderRequest(name: "C# Quick Start"));
        var folder = folderResp.Ok()!;
        var folderId = folder.Id!.Value;
        Console.WriteLine($"Created folder: {folder.Name} (id={folderId})");

        // ──── 3. Upload a file ──────────────────────────────────────────────
        var tempFile = Path.GetTempFileName() + ".txt";
        await File.WriteAllTextAsync(tempFile, "Hello from OpenFilz C# SDK!");

        try
        {
            using var fileStream = File.OpenRead(tempFile);
            var uploadResp = await documentApi.UploadDocument1Async(
                file: fileStream,
                allowDuplicateFileNames: true,
                parentFolderId: folderId.ToString());
            var uploaded = uploadResp.Ok()!;
            var fileId = uploaded.Id!.Value;
            Console.WriteLine($"Uploaded: {uploaded.Name} (id={fileId})");

            // ──── 4. List folder contents ───────────────────────────────────────
            var listResp = await folderApi.ListFolderAsync(folderId: folderId, onlyFiles: false, onlyFolders: false);
            var contents = listResp.Ok()!;
            Console.WriteLine($"Folder contains {contents.Count} item(s):");
            foreach (var item in contents)
            {
                Console.WriteLine($"  {item.Type}: {item.Name}");
            }

            // ──── 5. Get document info ──────────────────────────────────────────
            var infoResp = await documentApi.GetDocumentInfoAsync(fileId, withMetadata: true);
            var info = infoResp.Ok()!;
            Console.WriteLine($"Document: name={info.Name}, type={info.ContentType}");

            // ──── 6. Update metadata ────────────────────────────────────────────
            await documentApi.UpdateDocumentMetadataAsync(fileId,
                new UpdateMetadataRequest(
                    metadataToUpdate: new Dictionary<string, object>
                    {
                        { "project", "C# Demo" },
                        { "version", 1 }
                    }));
            Console.WriteLine("Metadata updated");

            // ──── 7. Download the file ──────────────────────────────────────────
            var downloadResp = await documentApi.DownloadDocumentAsync(fileId);
            var downloadedStream = downloadResp.Ok()!;
            Console.WriteLine("File downloaded");

            // ──── 8. Rename the file ────────────────────────────────────────────
            var renameResp = await fileApi.RenameFileAsync(fileId,
                new RenameRequest(newName: "csharp-renamed.txt"));
            var renamed = renameResp.Ok()!;
            Console.WriteLine($"Renamed to: {renamed.Name}");

            // ──── 9. Create target folder and copy ──────────────────────────────
            var targetFolderResp = await folderApi.CreateFolderAsync(
                new CreateFolderRequest(name: "C# Quick Start Target"));
            var targetFolderId = targetFolderResp.Ok()!.Id!.Value;

            var copyResp = await fileApi.CopyFilesAsync(
                new CopyRequest(
                    documentIds: new List<Guid> { fileId },
                    targetFolderId: targetFolderId));
            var copies = copyResp.Ok()!;
            var copyId = copies[0].CopyId!.Value;
            Console.WriteLine($"Copied file, new id={copyId}");

            // ──── 10. Move the original file ────────────────────────────────────
            await fileApi.MoveFilesAsync(
                new MoveRequest(
                    documentIds: new List<Guid> { fileId },
                    targetFolderId: targetFolderId,
                    allowDuplicateFileNames: true));
            Console.WriteLine("Moved file to target folder");

            // ──── 11. Toggle favorite ───────────────────────────────────────────
            var favResp = await favApi.ToggleFavoriteAsync(fileId);
            var isFavorite = favResp.Ok()!;
            Console.WriteLine($"Favorite toggled: {isFavorite}");

            // ──── 12. Dashboard statistics ──────────────────────────────────────
            var statsResp = await dashboardApi.GetDashboardStatisticsAsync();
            var stats = statsResp.Ok()!;
            Console.WriteLine($"Dashboard: files={stats.TotalFiles}, folders={stats.TotalFolders}");

            // ──── 13. Audit trail ───────────────────────────────────────────────
            var trailResp = await auditApi.GetAuditTrailAsync(fileId, sort: "DESC");
            var trail = trailResp.Ok()!;
            Console.WriteLine($"Audit trail: {trail.Count} entries");

            // ──── 14. Cleanup ───────────────────────────────────────────────────
            await favApi.RemoveFavoriteAsync(fileId);
            await fileApi.DeleteFilesAsync(
                new DeleteRequest(documentIds: new List<Guid> { fileId, copyId }));
            await folderApi.DeleteFoldersAsync(
                new DeleteRequest(documentIds: new List<Guid> { folderId, targetFolderId }));
            Console.WriteLine("Cleanup complete");
        }
        finally
        {
            if (File.Exists(tempFile)) File.Delete(tempFile);
        }
    }
}

File: openfilz-sdk/openfilz-sdk-java/samples/QuickStart.java
import org.openfilz.sdk.ApiClient;
import org.openfilz.sdk.ApiException;
import org.openfilz.sdk.Configuration;
import org.openfilz.sdk.api.*;
import org.openfilz.sdk.model.*;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileReader;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.List;
import java.util.Map;
import java.util.UUID;

/**
 * OpenFilz Java SDK Quick Start
 *
 * This sample demonstrates the core document management operations:
 * - Folder creation
 * - File upload, download, rename, move, copy, delete
 * - Favorites management
 * - Metadata updates
 * - Dashboard statistics
 * - Audit trail
 *
 * Prerequisites:
 *   - A running OpenFilz API instance (default: http://localhost:8081)
 *   - Maven dependency: org.openfilz:openfilz-sdk-java
 */
public class QuickStart {

    public static void main(String[] args) throws ApiException, IOException {
        // ──── 1. Configure the SDK client ────────────────────────────────
        ApiClient client = Configuration.getDefaultApiClient();
        client.setBasePath("http://localhost:8081");

        // For authenticated instances, configure Bearer token:
        // HttpBearerAuth auth = (HttpBearerAuth) client.getAuthentication("keycloak_auth");
        // auth.setBearerToken("your-access-token");

        FolderControllerApi folderApi = new FolderControllerApi(client);
        DocumentControllerApi documentApi = new DocumentControllerApi(client);
        FileControllerApi fileApi = new FileControllerApi(client);
        FavoritesApi favApi = new FavoritesApi(client);
        DashboardApi dashboardApi = new DashboardApi(client);
        AuditControllerApi auditApi = new AuditControllerApi(client);

        // ──── 2. Create a folder ─────────────────────────────────────────
        CreateFolderRequest folderRequest = new CreateFolderRequest();
        folderRequest.setName("Quick Start Demo");
        folderRequest.setParentId(null); // root folder

        FolderResponse folder = folderApi.createFolder(folderRequest);
        UUID folderId = folder.getId();
        System.out.println("Created folder: " + folder.getName() + " (id=" + folderId + ")");

        // ──── 3. Upload a file into the folder ───────────────────────────
        Path tempFile = Files.createTempFile("demo-", ".txt");
        Files.writeString(tempFile, "Hello from OpenFilz SDK!");

        UploadResponse uploaded = documentApi.uploadDocument1(
                tempFile.toFile(),
                true,                       // allowDuplicateFileNames
                folderId.toString(),         // parentFolderId
                null                         // metadata JSON (optional)
        );
        UUID fileId = uploaded.getId();
        System.out.println("Uploaded: " + uploaded.getName() + " (id=" + fileId + ")");

        // ──── 4. List folder contents ────────────────────────────────────
        List<FolderElementInfo> contents = folderApi.listFolder(folderId, false, false);
        System.out.println("Folder contains " + contents.size() + " item(s):");
        for (FolderElementInfo item : contents) {
            System.out.println("  " + item.getType() + ": " + item.getName());
        }

        // ──── 5. Get document info ───────────────────────────────────────
        DocumentInfo info = documentApi.getDocumentInfo(fileId, true);
        System.out.println("Document info: name=" + info.getName()
                + ", contentType=" + info.getContentType()
                + ", size=" + info.getSize());

        // ──── 6. Update metadata ─────────────────────────────────────────
        UpdateMetadataRequest metadataReq = new UpdateMetadataRequest();
        metadataReq.setMetadataToUpdate(Map.of(
                "project", "Quick Start",
                "classification", "demo"
        ));
        documentApi.updateDocumentMetadata(fileId, metadataReq);
        System.out.println("Metadata updated");

        // ──── 7. Download the file ───────────────────────────────────────
        // Stream-based reading — avoids loading entire file into memory
        File downloaded = documentApi.downloadDocument(fileId);
        try (BufferedReader reader = new BufferedReader(new FileReader(downloaded))) {
            String line;
            while ((line = reader.readLine()) != null) {
                System.out.println("Downloaded content: " + line);
            }
        }

        // ──── 8. Rename the file ─────────────────────────────────────────
        RenameRequest rename = new RenameRequest();
        rename.setNewName("renamed-demo.txt");
        ElementInfo renamed = fileApi.renameFile(fileId, rename);
        System.out.println("Renamed to: " + renamed.getName());

        // ──── 9. Create a target folder and copy the file ────────────────
        CreateFolderRequest targetReq = new CreateFolderRequest();
        targetReq.setName("Quick Start Target");
        FolderResponse targetFolder = folderApi.createFolder(targetReq);
        UUID targetFolderId = targetFolder.getId();

        CopyRequest copyReq = new CopyRequest();
        copyReq.setDocumentIds(List.of(fileId));
        copyReq.setTargetFolderId(targetFolderId);
        List<CopyResponse> copies = fileApi.copyFiles(copyReq);
        UUID copyId = copies.getFirst().getCopyId();
        System.out.println("Copied file, new id=" + copyId);

        // ──── 10. Move the original file ─────────────────────────────────
        MoveRequest moveReq = new MoveRequest();
        moveReq.setDocumentIds(List.of(fileId));
        moveReq.setTargetFolderId(targetFolderId);
        moveReq.setAllowDuplicateFileNames(true);
        fileApi.moveFiles(moveReq);
        System.out.println("Moved file to target folder");

        // ──── 11. Toggle favorite ────────────────────────────────────────
        Boolean isFavorite = favApi.toggleFavorite(fileId);
        System.out.println("Favorite toggled: " + isFavorite);

        // ──── 12. Dashboard statistics ───────────────────────────────────
        DashboardStatisticsResponse stats = dashboardApi.getDashboardStatistics();
        System.out.println("Dashboard: totalFiles=" + stats.getTotalFiles()
                + ", totalFolders=" + stats.getTotalFolders());

        // ──── 13. Audit trail ────────────────────────────────────────────
        List<AuditLog> trail = auditApi.getAuditTrail(fileId, "DESC");
        System.out.println("Audit trail (" + trail.size() + " entries):");
        for (AuditLog entry : trail) {
            System.out.println("  " + entry.getAction() + " at " + entry.getTimestamp());
        }

        // ──── 14. Cleanup ────────────────────────────────────────────────
        favApi.removeFavorite(fileId);
        DeleteRequest deleteFiles = new DeleteRequest();
        deleteFiles.setDocumentIds(List.of(fileId, copyId));
        fileApi.deleteFiles(deleteFiles);

        DeleteRequest deleteFolders = new DeleteRequest();
        deleteFolders.setDocumentIds(List.of(folderId, targetFolderId));
        folderApi.deleteFolders(deleteFolders);
        System.out.println("Cleanup complete");

        Files.deleteIfExists(tempFile);
    }
}
